cursive-core/src/logger.rs

```diff
@@ -2,14 +2,107 @@
+///
+/// # Examples
+///
+/// Set log levels from env vars
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// CursiveLogger::new()
+///     .with_env()
+///     .init();
+/// ```
+///
+/// Set log levels explicitly.
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// # use log::LevelFilter;
+/// CursiveLogger::new()
+///     .with_int_filter_level(LevelFilter::Warn)
+///     .with_ext_filter_level(LevelFilter::Debug)
+///     .init();
+/// ```
+///
+/// Set log queue size.
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// CursiveLogger::new()
+///     .with_log_size(10_000)
+///     .init();
+/// ```
+pub struct CursiveLogger {
+    // Log filter level for log messages from within cursive
+    int_filter_level: log::LevelFilter,
+    // Log filter level for log messages from sources outside of cursive
+    ext_filter_level: log::LevelFilter,
+    // Size of log queue
+    log_size: usize,
+}
+
+impl CursiveLogger {
+    /// Creates a new CursiveLogger with default log filter levels of `log::LevelFilter::Trace`.
+    /// Remember to call `init()` to install with `log` backend.
+    pub fn new() -> Self {
+        CursiveLogger {
+            int_filter_level: log::LevelFilter::Trace,
+            ext_filter_level: log::LevelFilter::Trace,
+            log_size: 1000,
+        }
+    }
+
+    /// Sets the internal log filter level.
+    pub fn with_int_filter_level(mut self, level: log::LevelFilter) -> Self {
+        self.int_filter_level = level;
+        self
+    }
+
+    /// Sets the external log filter level.
+    pub fn with_ext_filter_level(mut self, level: log::LevelFilter) -> Self {
+        self.ext_filter_level = level;
+        self
+    }
 
-static LOGGER: CursiveLogger = CursiveLogger;
+    /// Sets log filter levels based on environment variables `RUST_LOG` and `CURSIVE_LOG`.
+    /// If `RUST_LOG` is set, then both internal and external log levels are set to match.
+    /// If `CURSIVE_LOG` is set, then the internal log level is set to match with precedence over
+    /// `RUST_LOG`.
+    pub fn with_env(mut self) -> Self {
+        if let Ok(rust_log) = std::env::var("RUST_LOG") {
+            if let Ok(filter_level) = log::LevelFilter::from_str(&rust_log) {
+                self.int_filter_level = filter_level;
+                self.ext_filter_level = filter_level;
+            }
+        }
+        if let Ok(cursive_log) = std::env::var("CURSIVE_LOG") {
+            if let Ok(filter_level) = log::LevelFilter::from_str(&cursive_log) {
+                self.int_filter_level = filter_level;
+            }
+        }
+        self
+    }
+
+    /// Sets the size of the log queue
+    pub fn with_log_size(mut self, log_size: usize) -> Self {
+        self.log_size = log_size;
+        self
+    }
+
+    /// Installs the logger with log. Calling twice will panic.
+    pub fn init(self) {
+        reserve_logs(self.log_size);
+        log::set_logger(Box::leak(Box::new(self))).unwrap();
```

Comment:
We may want to use `set_boxed_logger`, which hides the leaking away so it feels less dirty (even though it's really the same):
https://docs.rs/log/latest/log/fn.set_boxed_logger.html

I guess a "leak-free" version would keep CursiveLogger zero-sized and rely on a global/singleton state instead, but that may be more complex, with quite limited benefits (prevents leaks if `init()` is called repeatedly).

Reply:
Yeah, I originally had used `set_boxed_logger` but that's hidden behind the `std` feature flag for `log` and I didn't want to change the dependencies unless I had to. 

As written, calling `init()` multiple times panics anyways, so I don't think we have to worry about leaks. Having everything kept in global state might not be a bad idea since we shouldn't expect anyone to create multiple `CursiveLogger` objects anyways.

I made another branch for this: [cursivelogger_global](https://github.com/blaisdellma/cursive/blob/8010b2c071455f5ffab97d485b62833a94a680e0/cursive-core/src/logger.rs). It is a bit more complicated and I don't know if I like the API ergonomics. One plus though is being able to change the log filter levels after initialization. I don't know if that's worth it.

Reply:
Why the `Mutex<Cell<...>>` in that branch? Could a simple mutex work too?

Reply:
I think I like this global branch! No feature change for the log dependency, uses lazy_static that we were already using...

Reply:
Yeah, both work, you're right. I don't know what I was thinking with `Mutex<Cell<...>>`. Since `LOGS` is a `Mutex` anyway I'm not sure there's much of a speed difference.

Reply:
I created a [PR for cursivelogger_global](https://github.com/blaisdellma/cursive/pull/1). There is a tiny issue with `set_log_size` though.